File: android/appmanager/src/main/java/com/lmig/uscm/digital/appplatform/appmanager/FeatureManager.kt
package com.lmig.uscm.digital.appplatform.appmanager

import android.content.Context
import android.content.Intent

/**
 * Feature Manager
 *
 * Class representing the Feature Manager
 *
 * @param [features] A [MutableList] of [Feature] modules used in the app
 * @param [defaultFeatureName] The name of the default (initial) [Feature] to be displayed
 * @param [context] The current Android [Context]
 *
 * @property [currentFeatureName] The name of the current [Feature] displayed
 */
class FeatureManager(val features: MutableList<Feature>, val defaultFeatureName: String, val context: Context) {

    var currentFeatureName = defaultFeatureName

    /**
     * Updates a feature with new config based on an updated [Navigation] object
     *
     * @param [feature] The [Feature] to be updated
     * @param [nav] A [Navigation] object containing the updates config data
     */
    fun updateFeature(feature: Feature, nav: Navigation){
        if (feature.config?.viewConfig?.config is ReactNativeViewConfig) {

            val conf = (feature.config?.viewConfig as ViewConfig).config as ReactNativeViewConfig

            val navData = nav.data as ReactNav

            // Edit the values of the current feature
            conf.moduleName = navData.moduleName
            conf.initialProperties = navData.initialProperties
            conf.launchOptions = navData.launchOptions

            feature.setup()
        }
    }

    /**
     * Displays a feature
     *
     * @param [feature] The [Feature] to be displayed
     */
    fun display(feature: Feature){
        // This should be part of feature setup, if the intention is to always run on the App's Context
        feature.intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        
        currentFeatureName = feature.name
        context.startActivity(feature.intent)
    }

    /**
     * Displays a feature with animation
     *
     * @param [feature] The [Feature] to be displayed
     * @param [animation] The animation type to be used
     */
    fun displayWithAnimation(feature: Feature, animation: String?){
        //TODO: Add animation
        currentFeatureName = feature.name
        context.startActivity(feature.intent)
    }

    /**
     * Merges updated [FeatureConfig] objects into the existing collection of [Feature] objects
     *
     * @param [featureConfig] An updated [MutableList] of [FeatureConfig] objects to be merged
     */
    fun mergeConfig(featureConfig: MutableList<FeatureConfig>){
        for (feature in features) {
            featureConfig.find { it.name == feature.name }?.let { featureConfig ->
                feature.config = featureConfig
            }
        }
    }

    /**
     * Merges updated [ViewConfig] objects into the existing collection of [ViewConfig] objects
     *
     * @param [viewConfig] An updated [MutableList] of [ViewConfig] objects to be merged
     */
    fun setupViews(viewConfig: MutableList<ViewConfig>){
        for (feature in features) {
            viewConfig.find { it.name == feature.config?.name }?.let { viewConfig ->
                feature.config?.viewConfig = viewConfig
            }
            feature.setup()
        }
    }

    /**
     * Gets the default [Feature] name
     *
     * @return The default [Feature] name
     */
    fun getDefaultFeature(): Feature? {
        return features.find { it.name == defaultFeatureName }
    }

    /**
     * Gets a [Feature] by name
     *
     * @param [name] The name of the feature to be returned
     *
     * @return A [Feature] with a matching name
     */
    fun getFeatureByName(name: String): Feature? {
        return features.find { it.name == name }
    }

}

File: ios/AppManager/FeatureManager.swift
//
//  ModuleManager.swift
//  AppManager
//
//  Copyright © 2017 Liberty Mutual. All rights reserved.
//

//// Manages all app features
open class FeatureManager {
    weak var appDelegate: UIApplicationDelegate?

    public var features: [Feature]
    public var defaultFeature: String
    public var currentFeatureName: String

    /**
     Initializer for app with multiple features
     
     - parameter features: Array of Features
     - parameter defaultFeature: Name of the feature that should be default
     */
    public init (features: [Feature], defaultFeature: String) {
        self.appDelegate = UIApplication.shared.delegate
        self.features = features
        self.defaultFeature = defaultFeature
        self.currentFeatureName = defaultFeature
    }

    /**
     Initializer for app with single feature
     
     - parameter feature: Single feature */
    public init (feature: Feature) {
        self.appDelegate = UIApplication.shared.delegate
        self.features = []
        self.features.append(feature)
        self.defaultFeature = feature.name
        self.currentFeatureName = feature.name
    }

    /**
     Updates feature with new navigational params
     
     - parameter feature: `Feature` to be updated
     - parameter nav: `Navigation` object used for updating Feature's view config
     */
    public func updateFeature(feature: Feature, nav: Navigation) {
        if feature.config?.viewConfig?.config is ReactNativeViewConfig {
            if var conf = feature.config?.viewConfig?.config as? ReactNativeViewConfig,
                let navData = nav.data as? ReactNav {
                conf.moduleName = navData.moduleName
                conf.initialProperties = navData.initialProperties
                conf.launchOptions = navData.launchOptions

                feature.config?.viewConfig?.config = conf
                feature.navigate()
            }
        }

        if feature.config?.viewConfig?.config is CordovaViewConfig {
            if var conf = feature.config?.viewConfig?.config as? CordovaViewConfig,
                let navData = nav.data as? CordovaNav {
                if let urlString = navData.url {
                    if let url = URL(string: urlString) {
                        conf.url = url
                        feature.config?.viewConfig?.config  = conf
                        feature.navigate()
                    }
                }
            }
        }
    }

    /**
     Takes the view controller from passed feature and displays it via app delegate
     
     - parameter feature: Feature to be displayed
     */
    public func display(feature: Feature) {
        self.appDelegate?.window??.rootViewController = feature.view?.viewController
        self.appDelegate?.window??.makeKeyAndVisible()
        self.currentFeatureName = feature.name
    }

    // TODO: Implement animation types
    public func displayWithAnimation(feature: Feature, animation: String?) {
        guard let window = self.appDelegate?.window else {
            return
        }

        UIView.transition(with: window!, duration: 0.3, options: .transitionCrossDissolve, animations: {
            window!.rootViewController = feature.view?.viewController
        }, completion: { _ in })

        self.currentFeatureName = feature.name
    }

    /**
     Merges the `[FeatureConfig]` with the local instance
     
     - parameter featureConfig: [FeatureConfig] for merging
     */
    public func mergeConfig(featureConfig: [FeatureConfig]) {
        for feature in self.features where feature.config == nil {
            feature.config = featureConfig.first(where: { $0.name == feature.name })
        }
    }

    /**
     Setups the `ViewConfig` for all features. Uses the name to match the config.
     Initializes the `View` for the feature
     
     - parameter viewConfig: [ViewConfig] used for setting up to all views
     */
    public func setupViews(viewConfig: [ViewConfig]) {
        for feature in self.features {
            if let vc = feature.config?.viewConfig {
                if vc.config == nil {
                    feature.config?.viewConfig = viewConfig.first(where: { $0.name == feature.config?.name })
                }
            }

            do {
                try feature.setup()
            } catch let error {
                print(error.localizedDescription)
            }
        }
    }

    /**
     Gets the default `Feature` from the instance
     
     - returns: Feature if found
     */
    public func getDefaultFeature() -> Feature? {
        for feature in self.features where feature.name == self.defaultFeature {
            return feature
        }
        return nil
    }

    /**
     Gets the `Feature` for given name
     
     - parameter name: Feature name string
     - returns: Feature if found
     */
    public func getFeatureByName(name: String) -> Feature? {
        for feature in self.features where feature.name == name {
            return feature
        }
        return nil
    }
}
